iwyu.cc

```diff
@@ -4138,6 +4093,14 @@ class IwyuAstConsumer
           ExtractProvidedTypeComponents(resugar_map));
     }
 
+    const auto [is_provided, comment] =
+        IsProvidedTypeComponent(current_ast_node());
+    // Don't call ReportTypeUse here, because scanning of template instantiation
+    // internals should be avoided: it is allowed not to provide some of
+    // template args.
+    if (is_provided)
```

Comment:
The testcase looks right, but I don't understand why this isn't `!is_provided`...?

Is it because a template specialization is "definition-side" and any type that's provided by the base template also needs to be reported here?

Reply:
`IsProvidedTypeComponent` function is used to determine types provided by the type alias or function author on the alias or function definition side. So, this code handles occurrences of (written) template spec types in that definitions and not on any use site of the alias or the function. For example:
```cpp
typedef Tpl<Class> Alias;
```
If the `Alias` author decided to provide the underlying type, IWYU should report `Tpl` as well as `Class` when traversing `Alias` definition. `VisitTemplateSpecializationType` should do that for `Tpl` and `VisitTagType` should do the same for `Class`.

Reply:
The problem was in that such an analysis was performed not only on type alias definitions (and probably fn declarations) _as-written_ but also on implicitly generated code (inside implicit template instantiations) which is handled by `InstantiatedTemplateVisitor`.

Reply:
Thanks, I think that makes sense -- so `VisitTemplateSpecializationType` hits the template instantiation/expansion/specialization and because it's part of a typedef/alias/function decl, the custom provision rule hits? I wonder if this affects any of the label:template-spill issues we had before. I happen to have a stack of testcases, I'll run them over this patch just to see if there's any coverage missing.

Reply:
Also, I feel like this bug might be related: https://github.com/include-what-you-use/include-what-you-use/issues/1349 -- but I can imagine there's more subtleties there.

Reply:
> I wonder if this affects any of the label:template-spill issues we had before. I happen to have a stack of testcases, I'll run them over this patch just to see if there's any coverage missing.

No surprises here, everything still works as before (which I think/hope means that my 10 reduced examples are covered by the test suite :crossed_fingers:).

Reply:
> Thanks, I think that makes sense -- so VisitTemplateSpecializationType hits the template instantiation/expansion/specialization and because it's part of a typedef/alias/function decl, the custom provision rule hits?

The problem appeared when `VisitTemplateSpecializationType` hit a template specialization type which is a part of a typedef/alias/function decl _which in turn is a part of an implicit instantiation of another template_.

I think this case is very specific, and users hardly faced it. The change is rather conceptual and is aimed to enable possibilities of further refactorings and improvements.

Reply:
> * `WithAlias<int, TplIndirectStruct2<int>> wa;` is the implicit instantiation of another template
> * `using Type = TplIndirectStruct2<T1>;` is the alias declaration
> * `TplIndirectStruct2<T1>` is the template specialization type

Yes, correct.